scripts/java/hilla-scripts/src/main/java/com/vaadin/hilla/scripts/FlowPackageJsonUpdater.java

```diff
@@ -0,0 +1,154 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.vaadin.hilla.scripts;
+
+import com.fasterxml.jackson.core.JsonPointer;
+import com.fasterxml.jackson.core.util.DefaultIndenter;
+import com.fasterxml.jackson.core.util.DefaultPrettyPrinter;
+import com.fasterxml.jackson.core.util.Separators;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.vaadin.flow.internal.JacksonUtils;
+import com.vaadin.flow.server.frontend.FileIOUtils;
+import com.vaadin.flow.server.frontend.NodeUpdater;
+import org.apache.commons.io.IOUtils;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.stream.Stream;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Utility for reading "package.json" Flow resources and applying changes to the
+ * Hilla repository when necessary.
+ * <p>
+ * For internal use only. May be renamed or removed in a future release.
+ */
+public class FlowPackageJsonUpdater {
+    private static final String FRONTEND_RESOURCES_PATH = NodeUpdater.class
+            .getPackage().getName().replace('.', '/');
+
+    private static final JsonPointer DEPENDENCIES = JsonPointer
+            .compile("/dependencies");
+    private static final JsonPointer DEV_DEPENDENCIES = JsonPointer
+            .compile("/devDependencies");
+
+    private Path packageJsonFile;
+    private ObjectNode tree;
+
+    private FlowPackageJsonUpdater(Path packageJsonFile) throws IOException {
+        this.packageJsonFile = packageJsonFile;
+        var packageJsonFileContents = Files.readString(this.packageJsonFile);
+        tree = JacksonUtils.readTree(packageJsonFileContents);
+    }
+
+    private void applyFlowPackageJson(String id) {
+        if (logger().isDebugEnabled()) {
+            logger().debug("Applying Flow {} package.json.", id);
+        }
+        var name = String.join("/", FRONTEND_RESOURCES_PATH, "dependencies", id,
+                "package.json");
+        var resource = getClass().getClassLoader().getResource(name);
+        try {
+            assert resource != null;
+            var flowPackageJsonTree = JacksonUtils.readTree(
+                    IOUtils.toString(resource, StandardCharsets.UTF_8));
+            Stream.of(DEPENDENCIES, DEV_DEPENDENCIES).forEach((section) -> {
+                // NOTE: Some dependencies are devDependencies in Hilla,
+                // such as "react". Hence we update at "devDependencies".
+                updateTreeAt(DEV_DEPENDENCIES, flowPackageJsonTree.at(section));
+            });
```

Comment:
Does dev dependencies to be also updated here? It looks like you're trying to update dev dependencies to themselves, or I'm wrong?

Reply:
The target category is always “devDepencencies”, cause the repo's root `package.json` uses only that. The source categories are both “dependencies” and “devDependencies”, as “package.json” resouces in Flow use both.